packages/upgrader/src/upgrade.ts

```diff
@@ -43,31 +43,110 @@ import { waitForControllerDeployment } from "./readiness";
 
 const CONTROLLER_IMAGE_DEFAULT = `opstrace/controller:${BUILD_INFO.VERSION_STRING}`;
 
-//
-// Set the controller deployment image version to the one defined in buildinfo.
-// Returns boolean to indicate controller deployment rollout initiated or not.
-export function* upgradeControllerDeployment(config: {
-  opstraceClusterName: string;
-  kubeConfig: KubeConfig;
-}): Generator<Effect, boolean, State> {
-  // Exit if controller deployment does not exist.
+// Checks if the  cliMetadata.allCLIVersions image tag of the Opstrace
+// controller deployment matches the one defined in buildinfo. Older Opstrace
+// versions will have an empty cliMetadata field, for these, check the Opstrace
+// controller deployment image and check it matches buildinfo.
+export function* opstraceInstanceRequiresUpgrade(): Generator<
+  Effect,
+  boolean,
+  any
+> {
+  const opstraceControllerConfig: LatestControllerConfigType = yield call(
+    getOpstraceControllerConfigMap
+  );
+
+  const opstraceDeployment: Deployment = yield call(
+    getOpstraceControllerDeployment
+  );
+
+  const installedControllerVersion =
+    opstraceDeployment.spec.spec?.template.spec?.containers[0].image;
+
+  if (opstraceControllerConfig.cliMetadata.allCLIVersions.length > 1) {
+    // The latest version the instance was upgraded to was pushed to
+    // allCLIVersions in `upgradeControllerConfigMap` function.
+    const idx = opstraceControllerConfig.cliMetadata.allCLIVersions.length - 1;
+    const lastCLIVersion =
+      opstraceControllerConfig.cliMetadata.allCLIVersions[idx].version;
+
+    const installedControllerImageTag = installedControllerVersion
+      ?.split(":")
+      .pop();
+    if (lastCLIVersion !== installedControllerImageTag) {
+      die(
+        `found Opstrace version mismatch, CLI version ${lastCLIVersion} does not match installed controller version ${installedControllerImageTag}`
```

Comment:
:+1: 

In this error message, "CLI version" sounds like "the version of the CLI that's currently being executed".

Maybe let's expose an implementation detail here and make sure the error message is "correct". This is like an `assert()` in code where we really think this invariant is never violated anyway. So when this blows up, the error message may be optimized for debuggability / ugly.

Let's maybe say "... lastCLIVersion in the controller config is ${lastCLIVersion}, and does not match ..."